parsers/NA.py

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env python3
+
+"""Parser for Namibia."""
+
+import arrow
+from bs4 import BeautifulSoup
+from .lib.validation import validate
+from PIL import Image
+from pytesseract import image_to_string
+import re
+import requests
+
+
+generation_link = 'http://www.nampower.com.na/gis/images/Gx.png'
+exchanges_link = 'http://www.nampower.com.na/gis/images/Imports_Exports.png'
+
+plant_mapping = {"Ruacana": "hydro",
+                 "Van Eck": "coal",
+                 "Paratus": "oil",
+                 "Anixas": "oil",
+                 "Solar": "solar",
+                 "Wind": "wind"
+                 }
+
+exchange_mapping = {"NA->ZA": "ESKOM",
+                    "NA->ZM": "ZESCO"
+                    }
+
+
+def get_text_from_image(link, expected_size, new_size, session=None):
+    """
+    Gets image from link and checks expected size vs actual.
+    Converts to black & white and enlarges to improve OCR accuracy.
+    Performs OCR using tesseract and returns a str.
+    """
+
+    s = session or requests.Session()
+    img = Image.open(s.get(link, stream=True).raw)
+
+    if img.size != expected_size:
+        print("Check Namibia Scada dashboard for {} changes.".format(link))
+
+    gray = img.convert('L')
+    gray_enlarged = gray.resize(new_size, Image.LANCZOS)
+    text = image_to_string(gray_enlarged, lang='eng')
+
+    return text
+
+
+def data_processor(text):
+    """
+    Takes text produced from OCR and extracts production.
+    Returns a dictionary.
+    """
+
+    production = {}
+    for k in plant_mapping.keys():
+        pattern = re.escape(k) + r": (\d+\.\d\d)"
+        try:
+            val = re.search(pattern, text).group(1)
+            production[plant_mapping[k]] = production.get(plant_mapping[k], 0.0)+ float(val)
+        except (AttributeError, ValueError) as e:
+            production[plant_mapping[k]] = None
+
+    return production
+
+
+def fetch_production(country_code = 'NA', session=None):
+    """
+    Requests the last known production mix (in MW) of a given country
+    Arguments:
+    country_code (optional) -- used in case a parser is able to fetch multiple countries
+    session (optional)      -- request session passed in order to re-use an existing session
+    Return:
+    A dictionary in the form:
+    {
+      'countryCode': 'FR',
+      'datetime': '2017-01-01T00:00:00Z',
+      'production': {
+          'biomass': 0.0,
+          'coal': 0.0,
+          'gas': 0.0,
+          'hydro': 0.0,
+          'nuclear': null,
+          'oil': 0.0,
+          'solar': 0.0,
+          'wind': 0.0,
+          'geothermal': 0.0,
+          'unknown': 0.0
+      },
+      'storage': {
+          'hydro': -10.0,
+      },
+      'source': 'mysource.com'
+    }
+    """
+
+    raw_text = get_text_from_image(session=None, link=generation_link, \
```

Comment:
Probably use `session` variable here and in fetch_exchange

Reply:
so session=session to avoid resetting it again?

Reply:
yeah, session=session will give get_text_from_image the same session object that fetch_production received